packages/flutter/lib/src/services/scribble.dart

```diff
@@ -0,0 +1,228 @@
+        final Rect bounds = _scribbleClients[elementIdentifier]?.bounds ?? Rect.zero;
+        return !(bounds == Rect.zero || bounds.hasNaN || bounds.isInfinite);
+      }).map((String elementIdentifier) {
+        final Rect bounds = _scribbleClients[elementIdentifier]!.bounds;
+        return <dynamic>[elementIdentifier, ...<dynamic>[bounds.left, bounds.top, bounds.width, bounds.height]];
+      }).toList();
+    } else if (method == 'Scribble.scribbleInteractionBegan') {
+      _scribbleInProgress = true;
+      return;
+    } else if (method == 'Scribble.scribbleInteractionFinished') {
+      _scribbleInProgress = false;
+      return;
+    }
+
+    final List<dynamic> args = methodCall.arguments as List<dynamic>;
+    switch (method) {
+      case 'Scribble.showToolbar':
+        _client!.showToolbar();
+        break;
+      case 'Scribble.insertTextPlaceholder':
+        _client!.insertTextPlaceholder(Size((args[1] as num).toDouble(), (args[2] as num).toDouble()));
+        break;
+      case 'Scribble.removeTextPlaceholder':
+        _client!.removeTextPlaceholder();
+        break;
+      default:
+        throw MissingPluginException();
+    }
+  }
+
+  /// Registers a [ScribbleClient] with [elementIdentifier] that can be focused
+  /// by the engine.
+  ///
+  /// For example, the registered [ScribbleClient] list is used to respond to
+  /// UIIndirectScribbleInteraction on an iPad.
+  static void registerScribbleElement(String elementIdentifier, ScribbleClient scribbleClient) {
+    _instance._scribbleClients[elementIdentifier] = scribbleClient;
+  }
+
+  /// Unregisters a [ScribbleClient] with [elementIdentifier].
+  static void unregisterScribbleElement(String elementIdentifier) {
+    _instance._scribbleClients.remove(elementIdentifier);
+  }
+
+  List<SelectionRect> _cachedSelectionRects = <SelectionRect>[];
+
+  /// Send the bounding boxes of the current selected glyphs in the client to
+  /// the platform's text input plugin.
+  ///
+  /// These are used by the engine during a UIDirectScribbleInteraction.
+  static void setSelectionRects(List<SelectionRect> selectionRects) {
+    if (!listEquals(_instance._cachedSelectionRects, selectionRects)) {
+      _instance._cachedSelectionRects = selectionRects;
+      _instance._channel.invokeMethod<void>(
+        'Scribble.setSelectionRects',
+        selectionRects.map((SelectionRect rect) {
+          return <num>[rect.bounds.left, rect.bounds.top, rect.bounds.width, rect.bounds.height, rect.position];
+        }).toList(),
+      );
+    }
+  }
+}
+
+/// An interface to receive focus from the engine.
+///
+/// This is currently only used to handle UIIndirectScribbleInteraction.
+mixin ScribbleClient {
+  /// A unique identifier for this element.
+  String get elementIdentifier;
+
+  /// Called by the engine when the [ScribbleClient] should receive focus.
+  ///
+  /// For example, this method is called during a UIIndirectScribbleInteraction.
+  void onScribbleFocus(Offset offset);
```

Comment:
The documentation didn't explain `offset`?

Reply:
No, but I'll add something about it.